README.md
[![Appveyor build Status](https://ci.appveyor.com/api/projects/status/github/KWB-R/qmra.db?branch=master&svg=true)](https://ci.appveyor.com/project/KWB-R/qmra-db/branch/master)
[![Travis build Status](https://travis-ci.org/KWB-R/qmra.db.svg?branch=master)](https://travis-ci.org/KWB-R/qmra.db)
[![codecov](https://codecov.io/github/KWB-R/qmra.db/branch/master/graphs/badge.svg)](https://codecov.io/github/KWB-R/qmra.db)
[![Project Status](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing)
[![CRAN_Status_Badge](https://www.r-pkg.org/badges/version/qmra.db)]()
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.3339321.svg)](https://doi.org/10.5281/zenodo.3339321)

# qmra.db

This package contains functions for exporting
tables of MS ACCESS database into .csv files.

## Installation

```r
###############################################################################
### 1 Install R Package Downloading and Installing from GitHub
###############################################################################

install.packages("remotes", repos = "https://cloud.r-project.org")

#Sys.setenv(GITHUB_PAT = "mysecret_access_token")
remotes::install_github("r-lib/remotes@18c7302637053faf21c5b025e1e9243962db1bdc")

###############################################################################
### 2 Install R Package qmra.db (with ACCESS DB)
###############################################################################

### 2.1 Specific Release ACCESS DB from our GitHub account, e.g.

#remotes::install_github("kwb-r/qmra.db@0.9.0")

### 2.2 the Latest Development Version 

remotes::install_github("kwb-r/qmra.db")  
```

## Documentation

Release: [https://kwb-r.github.io/qmra.db](https://kwb-r.github.io/qmra.db)

Development: [https://kwb-r.github.io/qmra.db/dev](https://kwb-r.github.io/qmra.db/dev)

## CHANGELOG Moved To [NEWS.md](https://github.com/KWB-R/qmra.db/blob/master/NEWS.md)

Formerly we used the **README.md** for tracking what has change between different versions. 

This changelog is now in [NEWS.md](https://github.com/KWB-R/qmra.db/blob/master/NEWS.md)


NEWS.md
# [qmra.db 0.10.0](https://github.com/KWB-R/qmra.db/releases/tag/v0.10.0) <small>2019-07-17</small>

Prepare for first release on [KWB community on Zenodo](https://zenodo.org/communities/kwb/):

* add vignette for checking database content and add new function 
import_database_content() for importing the content into R ([#5](https://github.com/KWB-R/qmra.db/issues/5)) 

* made qmra.db public and build documentation in gh-pages (and use Gitlab only 
as backup)


# [qmra.db 0.9.0](https://github.com/KWB-R/qmra.db/releases/tag/v0.9.0) <small>2019-02-21</small>

**BREAKING CHANGES**: ***due to converting repo as R package***

Moved from [/](https://github.com/KWB-R/qmra.db/tree/master/) into 
[/inst/database](https://github.com/KWB-R/qmra.db/tree/master/inst) directory 
(required for R Package):

* MS ACCESS database ([qmra-db.accdb](https://github.com/KWB-R/qmra.db/tree/master/inst/database)), 

* folder: [qmra-db_accdb](https://github.com/KWB-R/qmra.db/tree/master/inst/database/qmra-db_accdb) 
(exported MS ACCESS content as .csv)

* folder: [rawdata](https://github.com/KWB-R/qmra.db/tree/master/inst/database/rawdata)  (Excel files received from @PatrickSmeetsKWR and imported into MS ACCESS DB by @mrustl), see NEWS entries of prior qmra.db releases

* folder: [ui-workflow](https://github.com/KWB-R/qmra.db/tree/master/inst/database/ui-workflow)  
(flowchart related to development of QMRA webapp user interface by @mrustl)


R files: 

* Moved "db-initialGeneration.R" from [/](https://github.com/KWB-R/qmra.db/tree/master/) into [data-raw](https://github.com/KWB-R/qmra.db/tree/master/data-raw) folder


* Moved "db-export.R" from [/](https://github.com/KWB-R/qmra.db/tree/master/) into [R](https://github.com/KWB-R/qmra.db/tree/master/data-raw) folder. In addition 
added Rscript [.init_package.R](https://github.com/KWB-R/qmra.db/commit/dde0576ab57575135a1e16a81f752bf1bce05d98#diff-990bb284e3185198984dba48ab42ba2e)" describing workflow for preparing this 
R package.


tbl_inflow: 

* Use "log10_norm" as default for all data entries (instead of "uniform") in MS ACCESS DB  and export csv files ([#2](https://github.com/KWB-R/qmra.db/issues/2)), which is now 
implemented in R package [kwb.qmra](https://github.com/kwb-r/kwb.qmra) (see [issue #14](https://github.com/KWB-R/kwb.qmra/issues/14) and [issue #17](https://github.com/KWB-R/kwb.qmra/issues/17))


***Prepare public release (#1)***

* Convert repository into R package for improved documentation and 
  also offers possiblity to add Rmarkdown export of MS ACCESS DB     
  (feature asked by @daniel-wicke)
  
* Add authors: @wseis @PatrickSmeetsKWR and Aquanes as "funder"

* Added a `NEWS.md` file to track changes to the package. Based on README.md but renamed original names to Github usernames

* see https://style.tidyverse.org/news.html for writing a good `NEWS.md`


# [qmra.db 0.8.0](https://github.com/KWB-R/qmra.db/releases/tag/v0.8.0) <small>2018-11-12</small>

"rawdata" folder: 
* added @PatrickSmeetsKWR's changes made in EXCEL files "tbl_logremoval.xlsx" and "tbl_treatment.xlsx"(sent to @mrustl on 
2018-11-08 via email) in "rawdata" folder

tbl_treatment: 
* added TreatmentDescription: from @PatrickSmeetsKWR's "tbl_logremoval.xlsx" (sent on 2018-11-08 via email)
* renamed/added TreatmentNames: UV disinfection 20 mJ/cm2, drinking and UV disinfection 40 mJ/cm2, drinking

tbl_logRemoval
* deleted entries for TreatmentID 15 (RO) from WHO (2011) (Reference: 8), now default RO data are based on NSF/ANSI 419  validation (referenceID = 47)



# [qmra.db 0.7.1](https://github.com/KWB-R/qmra.db/releases/tag/v0.7.1) <small>2018-10-11</small>

tbl_treatment: 
* added TreatmentGroup "Wetland" 
* added TreatmentDescription" (from WHO 2017: http://www.who.int/water_sanitation_health/publications/drinking-water-quality-guidelines-4-including-1st-addendum/en/)
* deleted TreatmentID 6 -> "Membrane filtration (micro, ultra-, nanofiltration, reverse osmosis)"

tbl_logRemoval
* deleted entries for TreatmentID 6 from WHO (2011)

# [qmra.db 0.7.0](https://github.com/KWB-R/qmra.db/releases/tag/v0.7.0) <small>2018-10-09</small>

tbl_treatment: renamed treatment processes
* UV ->  UV disinfection, drinking water
* UV radiation -> UV disinfection, wastewater
* Ozone -> Ozonation, drinking water
* Ozonation -> Ozonation, wastewater
* deleted TreatmentID 29 (no further entries) 


tbl_references: 
* added data from @PatrickSmeetsKWR`s EXCEL files (tbl_logRemoval_181008.xlsx, 
tbl_inflow_181008.xlsx) 

tbl_waterSource
* renamed WaterSourceID 7 from "drinking water" to "groundwater" and adapted WaterSourceDescription field accordingly

tbl_ingestion
* added data from @PatrickSmeetsKWR`s EXCEL file (Book1.xlsx, renamed by @mrustl to tbl_ingestion.xlsx)
* added raw data (used for updating DB) to sub-folder: rawdata/tbl_ingestion.xlsx
* added WaterUseID 9 (-> domestic use, washing maschine)

tbl_inflow: 
* added data from @PatrickSmeetsKWR`s EXCEL file (tbl_inflow_181008.xlsx)

   + new columns: mean, alpha, beta, distribution,      
     PathogenInReference,	Notes   
   
   + added only data values for mean, alpha, beta in cases these are       either not to be assumed the same as min/max (i.e. for   
     alpha/beta) or not calcululated i.e. mean = (min+max)/2
     
   + distribution: is set to "uniform" (as default)
- added raw data (used for updating DB) to sub-folder: rawdata/tbl_inflow_181008.xlsx


tbl_logRemoval

* added corrected data (PathogenGroupID!) from @PatrickSmeetsKWR`s EXCEL file 
  (tbl_logRemoval_181008.xlsx):
   + new columns: mean, alpha, beta, distribution, Notes
   + no data values entered for mean, alpha, beta as these are either are always 
   to be assumed the same as min/max (i.e. alpha/beta) or calculated, i.e. 
   mean = (min+max)/2
   + distribution: is set to "uniform" (as default)
* added raw data (used for updating DB) to sub-folder: rawdata/tbl_logRemoval_181008.xlsx
* bugfix: corrected wrong ReferenceID for DEMOWARE reference 



R/db-export.R
#' Default Path to MS ACCESS Database 
#' @return path to MS ACCESS DB
#' @examples
#' default_access_db()
#' @export
default_access_db <- function() {
  
system.file("database/qmra-db.accdb", package = "qmra.db")  
  
}


#' Helper Function: Default Csv Directory
#'  
#' @param db_path path to MS Access DB (default: deufault_access_db())
#' @return csv directory but with all "." replaced with
#' with "_" (e.g. "qmra-db_accdb" instead of "qmra-db.accdb")
#' @importFrom stringr str_replace_all
#' @export
#' @examples 
#' db_path <- default_access_db()
#' default_csv_dir(db_path)
default_csv_dir <- function(db_path) {
  
stringr::str_replace_all(basename(db_path),
                         "\\.", 
                         "_")
}

#' Helper Function: Default Target Directory
#'  
#' @param db_path path to MS Access DB (default: deufault_access_db())
#' @param csv_dir csv directory (default: default_csv_dir()) 
#' @return subdirectory in MS ACCESS DB directory but with all "." replaced with
#' with "_" (e.g. "qmra-db_accdb" instead of "qmra-db.accdb")
#' @export
#' @examples 
#' db_path <- default_access_db()
#' default_target_dir(db_path)

default_target_dir <- function(db_path,
                               csv_dir = default_csv_dir(db_path)) {

  sprintf("%s/%s",
          dirname(db_path), 
          csv_dir
          )
  
}


#'  Dump MS ACCESS Database To Csv Files
#'  
#' @description for exporting tables and queries from MS ACCESS DB into 
#' .csv files (one for each table/query) into subfolder "qmra-db_accdb"
#' @param db_path path to MS Access DB (default: default_access_db())
#' @param target_dir directory. By default a new directory is created in the same 
#' directory as mdb resides in. The new directory has the same name as the 
#' database file with dots substituted with underscores (default: default_target_dir())
#' @param dbg debug information (default: TRUE)
#' @param ... additional arguments passed to kwb.db::dumpDatabase()
#' @importFrom kwb.utils safePath catAndRun
#' @importFrom kwb.db dumpDatabase
#' @export
dump_access_db <- function(db_path = default_access_db(),
                           target_dir = default_target_dir(db_path), 
                           dbg = TRUE,
                           ...) {


mdb_path <- kwb.utils::safePath(db_path)

kwb.utils::catAndRun(messageText = "Exporting MS Access Database content into .csv files",
                     expr = {
                       kwb.db::dumpDatabase(db = mdb_path, 
                                            pattern = c("^tbl|^qry"),
                                            target_dir = target_dir,
                                            ...
                                            )
                     }, 
                     dbg = dbg)
}

#'  Create Zip File with MS ACCESS Database And Exported CSV files 
#' @param db_path path to MS Access DB (default: default_access_db())
#' @param target_dir directory. By default a new zip file is created in the same 
#' directory as mdb resides in. The new directory has the same name as the 
#' database file but with file extension ".zip" (default: dirname(db_path))
#' @param dbg debug information (default: TRUE)
#' @importFrom fs dir_ls
#' @importFrom zip zip
#' @importFrom withr with_dir
#' @export

create_zipfile <- function(db_path,
                           target_dir = dirname(db_path),
                           dbg = TRUE) {
  

  dump_access_db(db_path, dbg = dbg)  
  
  source_dir <- dirname(db_path)
  
  zip_file <-  sprintf("%s.zip", basename(fs::path_ext_remove(db_path)))
  
  zip_path <- file.path(target_dir, zip_file)
  
  kwb.utils::catAndRun(messageText = sprintf("Creating zipfile '%s'", zip_path), 
                       expr = {
                         withr::with_dir(new = source_dir,
                                         code = {
                                           csv_paths <- fs::dir_ls(default_csv_dir(db_path),
                                                                   recursive = TRUE,
                                                                   type = "file",
                                                                   regexp = "[.]csv$")
                                           zip::zip(zipfile = zip_path,
                                                    files = c(basename(db_path),
                                                              csv_paths))
                                           }
                                         )},
                         dbg = dbg)
  
                         
      normalizePath(zip_path)
} 


R/.init_package.R

package  <- "qmra.db"

# Set the path to your new package
pkg_dir <- getwd()

stopifnot(basename(pkg_dir) == package)

# Create directory for R package
kwb.pkgbuild::create_pkg_dir(pkg_dir)

# Create a default package structure
withr::with_dir(pkg_dir, {kwb.pkgbuild::use_pkg_skeleton(package)})

kwb.orcid::get_kwb_orcids()

author <- list(
  name = "Michael Rustler",
  orcid = "0000-0001-9134-2871",
  url = "https://mrustl.de"
)

description <- list(
  name = package,
  title = "Database Backend for AQUANES QMRA Webapp",
  desc  = paste(
    "This package contains functions for exporting tables of MS ACCESS database", 
    "into .cSv files")
  )


kwb.pkgbuild::use_pkg(
  author,
  description,
  version = "0.9.0",
  stage = "maturing"
)


kwb.pkgbuild::use_gitlab_ci_pkgdown()


pkg_dependencies <- c("kwb.utils", "kwb.datetime", "kwb.event", "kwb.plot", 
                      "kwb.read", "kwb.monitoring", "kwb.logger")

sapply(pkg_dependencies, usethis::use_package)

sapply(sprintf("github::kwb-r/%s", pkg_dependencies), desc::desc_add_remotes)


R/check_db_content.R
#' Import Database Content From CSV Files
#'
#' @param csv_dir path to csv directory with database output (default: 
#'  system.file("database/qmra-db_accdb", package = "qmra.db"))
#' @param encoding encoding of csv files (default: "WINDOWS-1252")
#' @return list with multiple data.frames with database content
#' @export
#' @importFrom kwb.utils resolve
#' @importFrom rlang .data
#' @importFrom dplyr arrange left_join mutate n select 
#' @importFrom readr locale read_csv
#' @examples
#' \dontrun{
#' db_content <- import_database_content()
#' str(db_content, 1)
#' }
import_database_content <- function(
  csv_dir = system.file("database/qmra-db_accdb", package = "qmra.db"),
  encoding = "WINDOWS-1252") {
  
  paths_list <- list(
    csv_dir = system.file("database/qmra-db_accdb", package = "qmra.db"),
    dose_response = "<csv_dir>/tbl_doseResponse.csv", 
    guideline = "<csv_dir>/tbl_guideline.csv",  
    health = "<csv_dir>/tbl_health.csv", 
    inflow = "<csv_dir>/tbl_inflow.csv",
    ingestion = "<csv_dir>/tbl_ingestion.csv",
    log_removal = "<csv_dir>/tbl_logRemoval.csv", 
    pathogen = "<csv_dir>/tbl_pathogen.csv",
    pathogen_group = "<csv_dir>/tbl_pathogenGroup.csv",
    treatment = "<csv_dir>/tbl_treatment.csv",
    references = "<csv_dir>/tbl_reference.csv",
    water_source = "<csv_dir>/tbl_waterSource.csv"
  )
  
  paths <- kwb.utils::resolve(paths_list)
  
  
  read_exported_csv <- function (file) {
    readr::read_csv(file = file, 
                    locale = readr::locale(encoding = encoding)
                    )
  }
  
  dose_response <- read_exported_csv(file = paths$dose_response)
  guideline <- read_exported_csv(file = paths$guideline)
  health <- read_exported_csv(file = paths$health)
  inflow <- read_exported_csv(file = paths$inflow)
  ingestion <- read_exported_csv(file = paths$ingestion)
  log_removal <- read_exported_csv(file = paths$log_removal)
  pathogen <- read_exported_csv(file = paths$pathogen)
  pathogen_group <- read_exported_csv(file = paths$pathogen_group)
  treatment <- read_exported_csv(file = paths$treatment)
  references <- read_exported_csv(file = paths$references)
  water_source <- readr::read_csv(file = paths$water_source)
  
  references <-   references %>% 
    dplyr::mutate(ReferenceName = sprintf("[%s](%s)", 
                                          .data$ReferenceName, 
                                          .data$ReferenceLink)) %>% 
    dplyr::select(.data$ReferenceID, .data$ReferenceName)
  
  
  dose_response_metadata <- dose_response %>% 
    dplyr::left_join(pathogen) %>% 
    dplyr::left_join(pathogen_group) %>% 
    dplyr::left_join(references) %>% 
    dplyr::arrange(.data$PathogenGroup, 
                   .data$PathogenName, 
                   .data$ReferenceName) %>% 
    dplyr::mutate(row_id = 1:dplyr::n())
  
  guideline_metadata <- guideline %>% 
    dplyr::left_join(references) %>% 
    dplyr::arrange(.data$GuidelineName, 
                   .data$ReferenceName) %>% 
    dplyr::mutate(row_id = 1:dplyr::n())
  
  health_metadata <- health %>% 
    dplyr::left_join(pathogen) %>% 
    dplyr::left_join(pathogen_group) %>% 
    dplyr::left_join(references) %>% 
    dplyr::arrange(.data$PathogenGroup, 
                   .data$PathogenName, 
                   .data$ReferenceName) %>% 
    dplyr::mutate(row_id = 1:dplyr::n())
  
  
  inflow_metadata <- inflow %>% 
    dplyr::left_join(pathogen) %>% 
    dplyr::left_join(pathogen_group) %>% 
    dplyr::left_join(water_source) %>% 
    dplyr::left_join(references) %>% 
    dplyr::arrange(.data$WaterSourceName, 
                   .data$PathogenGroup, 
                   .data$PathogenName, 
                   .data$ReferenceName) %>% 
    dplyr::mutate(row_id = 1:dplyr::n())
  
  ingestion_metadata <- ingestion %>% 
    dplyr::left_join(references) %>% 
    dplyr::arrange(.data$WaterUseName, 
                   .data$ReferenceName, 
                   .data$ReferenceName) %>% 
    dplyr::mutate(row_id = 1:dplyr::n())
  
  treatment_metadata <- treatment %>% 
    dplyr::left_join(log_removal) %>% 
    dplyr::left_join(pathogen_group) %>% 
    dplyr::left_join(references) %>%
    dplyr::arrange(.data$TreatmentName,
                   .data$PathogenGroup, 
                   .data$ReferenceName) %>% 
    dplyr::mutate(row_id = 1:dplyr::n())
  
  list(dose_response = dose_response_metadata, 
       guideline = guideline_metadata,
       health = health_metadata, 
       inflow = inflow_metadata, 
       ingestion = ingestion_metadata,
       pathogen_group = pathogen_group %>%  
         dplyr::arrange(.data$PathogenGroup),
       treatment = treatment_metadata,
       water_source = water_source %>% 
         dplyr::arrange(.data$WaterSourceName))
}


db-scheme-for-kwb-qmra.R
devtools::install_github("kwb-r/kwb.qmra")
devtools::install_github("kwb-r/kwb.db")

library(dplyr)
library(kwb.qmra)

temp <- tempdir()
url <- "https://zenodo.org/record/159527/files/QMRA_OldFord.zip"
t_path <- file.path(temp, basename(url))
download.file(url, dest= t_path, mode="wb") 
unzip(t_path, exdir = temp)

scenarios_dir <- file.path(temp, "scenarios")

cat(sprintf("### Available scenarios:\n%s",
            paste(dir(scenarios_dir), collapse = "\n")))
    
    #### DEFINE DIRECTORY ################
    confDir <- dir(scenarios_dir,full.names = TRUE)[1]
    confDir
    config <- config_read(confDir)     
    

get_references <- function(df) {
  df %>% 
    dplyr::select(!!quo(ReferenceName), !!quo(ReferenceLink)) %>% 
    dplyr::filter(!is.na(!!quo(ReferenceName)))
}


config$doseresponse <- config$doseresponse %>%  
  dplyr::rename("ReferenceName" = "Reference", 
                "ReferenceLink" = "Link")

    
tbl_references <- get_references(config$inflow) %>%
    rbind(get_references(config$treatment$processes)) %>% 
    rbind(get_references(config$doseresponse )) %>%  
  rbind(get_references(config$health)) %>% 
  dplyr::group_by(ReferenceName, ReferenceLink) %>% 
  dplyr::summarise(n = n()) %>%  
  dplyr::ungroup() %>% 
  dplyr::mutate(ReferenceID = 1:n()) %>% 
  dplyr::select(ReferenceID, ReferenceName, ReferenceLink)
                  
        
tbl_pathogenGroups <- data.frame(PathogenGroup = unique(config$doseresponse$PathogenGroup), 
                         stringsAsFactors = FALSE) %>%
              dplyr::mutate(PathogenGroupID = 1:n()) %>% 
              dplyr::select(PathogenGroupID, PathogenGroup)

tbl_pathogens <- config$doseresponse[,1:3] %>% 
                 dplyr::left_join(tbl_pathogenGroups)


tbl_doseresponse <- config$doseresponse %>% 
  dplyr::left_join(tbl_references, by = c("ReferenceName", "ReferenceLink")) %>% 
  dplyr::left_join(tbl_pathogenGroups) 


tbl_health <- config$health %>% 
  dplyr::left_join(tbl_references, by = c("ReferenceName", "ReferenceLink")) %>% 
  dplyr::left_join(tbl_pathogens) %>% 
  dplyr::select(PathogenID, PathogenGroupID, ReferenceID, 
                infection_to_illness, dalys_per_case)

tbl_treatment_processes_per_patGroup_kwb <- config$treatment$processes %>% 
  dplyr::left_join(tbl_references, by = c("ReferenceName", "ReferenceLink")) %>% 
  dplyr::left_join(tbl_pathogenGroups)


tbl_watersource <- data.frame(WaterSourceID = c(1,2),
                              WaterSourceName = c("drinking water", "wastewater"), 
                              stringsAsFactors = FALSE)

tbl_processes <- tbl_treatment_processes_per_patGroup_kwb %>% 
    dplyr::group_by(TreatmentID, TreatmentName, TreatmentGroup, PathogenGroupID, 
                    PathogenGroup, ReferenceID, ReferenceName, ReferenceLink) %>% 
    dplyr::summarise(n = n()) %>% 
    dplyr::select(-n) %>% 
    dplyr::ungroup() %>%  
  dplyr::mutate(WaterSourceID = ifelse(grepl(pattern = 'WHO \\(2011\\)',
                                             ReferenceName), 
                                       1, 
                                       2)) %>% 
  dplyr::left_join(tbl_watersource)


tbl_treatment_processes <- tbl_processes %>% 
  dplyr::group_by(TreatmentID, TreatmentName, TreatmentGroup) %>% 
  dplyr::summarise(n = n()) %>% 
  dplyr::select(-n) %>%  
  dplyr::mutate(WaterSourceID = ifelse(TreatmentID < 16, 1, 2))



tbl_inflow <- config$inflow %>% 
  dplyr::group_by(PathogenGroup, simulate,  type, value,min,max, mode,  mean,
                  sd, meanlog, sdlog, ReferenceName,  ReferenceLink) %>%  
  dplyr::summarise(n = n()) %>% 
  dplyr::left_join(tbl_references, by = c("ReferenceName", "ReferenceLink")) %>% 
  dplyr::left_join(tbl_pathogenGroups) %>% 
  dplyr::select(-n) %>% 
  dplyr::ungroup() %>% 
  dplyr::mutate(WaterSourceID = 2) %>% 
  dplyr::select(PathogenGroupID, ReferenceID,  WaterSourceID, min, max)

path_db <- "C:/Users/mrustl.KWB/Documents/RProjects/kwb.qmra.accdb"

#file <- "C:/Users/mrustl.KWB/Documents/RProjects/qmra-db.RData"
#save.image(file = file)
#load(file)

table_names <- grep("^tbl", ls(), value = TRUE)

#table_names <- c("tbl_health", "tbl_inflow")

tables <- setNames(lapply(table_names, get), table_names)



for (table_name in table_names) {
  
  kwb.db::hsPutTable(
    path_db, tables[[table_name]], tbl = table_name, overwrite = TRUE
  )
}




db-initialGeneration.R
### R script for generating initial MS ACCESS DB structure based on Old Ford 
### public irrigation scenario
### (used for generating version v0.1/v0.2 of MS ACCESS DB)
### NOT REQUIRED FOR LATER VERSIONS oF MS ACCESS DB !!!!!!

if (FALSE) {
devtools::install_github("kwb-r/kwb.qmra")
devtools::install_github("kwb-r/kwb.db")

library(dplyr)
library(kwb.qmra)

temp <- tempdir()
url <- "https://zenodo.org/record/159527/files/QMRA_OldFord.zip"
t_path <- file.path(temp, basename(url))
download.file(url, dest= t_path, mode="wb") 
unzip(t_path, exdir = temp)

scenarios_dir <- file.path(temp, "scenarios")

cat(sprintf("### Available scenarios:\n%s",
            paste(dir(scenarios_dir), collapse = "\n")))
    
    #### DEFINE DIRECTORY ################
    confDir <- dir(scenarios_dir,full.names = TRUE)[1]
    confDir
    config <- config_read(confDir)     
    

get_references <- function(df) {
  df %>% 
    dplyr::select(!!quo(ReferenceName), !!quo(ReferenceLink)) %>% 
    dplyr::filter(!is.na(!!quo(ReferenceName)))
}


config$doseresponse <- config$doseresponse %>%  
  dplyr::rename("ReferenceName" = "Reference", 
                "ReferenceLink" = "Link")

    
tbl_references <- get_references(config$inflow) %>%
    rbind(get_references(config$treatment$processes)) %>% 
    rbind(get_references(config$doseresponse )) %>%  
  rbind(get_references(config$health)) %>% 
  dplyr::group_by(ReferenceName, ReferenceLink) %>% 
  dplyr::summarise(n = n()) %>%  
  dplyr::ungroup() %>% 
  dplyr::mutate(ReferenceID = 1:n()) %>% 
  dplyr::select(ReferenceID, ReferenceName, ReferenceLink)
                  
        
tbl_pathogenGroups <- data.frame(PathogenGroup = unique(config$doseresponse$PathogenGroup), 
                         stringsAsFactors = FALSE) %>%
              dplyr::mutate(PathogenGroupID = 1:n()) %>% 
              dplyr::select(PathogenGroupID, PathogenGroup)

tbl_pathogens <- config$doseresponse[,1:3] %>% 
                 dplyr::left_join(tbl_pathogenGroups)


tbl_doseresponse <- config$doseresponse %>% 
  dplyr::left_join(tbl_references, by = c("ReferenceName", "ReferenceLink")) %>% 
  dplyr::left_join(tbl_pathogenGroups) 


tbl_health <- config$health %>% 
  dplyr::left_join(tbl_references, by = c("ReferenceName", "ReferenceLink")) %>% 
  dplyr::left_join(tbl_pathogens) %>% 
  dplyr::select(PathogenID, PathogenGroupID, ReferenceID, 
                infection_to_illness, dalys_per_case)

tbl_treatment_processes_per_patGroup_kwb <- config$treatment$processes %>% 
  dplyr::left_join(tbl_references, by = c("ReferenceName", "ReferenceLink")) %>% 
  dplyr::left_join(tbl_pathogenGroups)


tbl_watersource <- data.frame(WaterSourceID = c(1,2),
                              WaterSourceName = c("drinking water", "wastewater"), 
                              stringsAsFactors = FALSE)

tbl_processes <- tbl_treatment_processes_per_patGroup_kwb %>% 
    dplyr::group_by(TreatmentID, TreatmentName, TreatmentGroup, PathogenGroupID, 
                    PathogenGroup, ReferenceID, ReferenceName, ReferenceLink) %>% 
    dplyr::summarise(n = n()) %>% 
    dplyr::select(-n) %>% 
    dplyr::ungroup() %>%  
  dplyr::mutate(WaterSourceID = ifelse(grepl(pattern = 'WHO \\(2011\\)',
                                             ReferenceName), 
                                       1, 
                                       2)) %>% 
  dplyr::left_join(tbl_watersource)


tbl_treatment_processes <- tbl_processes %>% 
  dplyr::group_by(TreatmentID, TreatmentName, TreatmentGroup) %>% 
  dplyr::summarise(n = n()) %>% 
  dplyr::select(-n) %>%  
  dplyr::mutate(WaterSourceID = ifelse(TreatmentID < 16, 1, 2))


# 
# tbl_inflow <- config$inflow %>% 
#   dplyr::group_by(PathogenName, PathogenGroup, simulate,  type, value,min,max, mode,  mean,
#                   sd, meanlog, sdlog, ReferenceName,  ReferenceLink) %>%  
#   dplyr::summarise(n = n()) %>% 
#   dplyr::left_join(tbl_references, by = c("ReferenceName", "ReferenceLink")) %>% 
#   dplyr::left_join(tbl_pathogenGroups) %>% 
#   dplyr::select(-n) %>% 
#   dplyr::ungroup() %>% 
#   dplyr::mutate(WaterSourceID = 2) %>% 
#   dplyr::select(PathogenGroupID, ReferenceID,  WaterSourceID, min, max)


tbl_inflow <- config$inflow %>% 
  dplyr::filter(!is.na(ReferenceName)) %>% 
  dplyr::select(PathogenName, PathogenGroup, min, max, 
                ReferenceName,  ReferenceLink) %>% 
  dplyr::left_join(tbl_references, by = c("ReferenceName", "ReferenceLink")) %>% 
  dplyr::left_join(tbl_pathogens) %>% 
  dplyr::mutate(WaterSourceID = 2) %>% 
  dplyr::select(PathogenID, PathogenGroupID, ReferenceID,  WaterSourceID, min, max)

path_db <- "C:/Users/mrustl.KWB/Desktop/QMRA_workflow/kwb.qmra-db/kwb.qmra.accdb"

#file <- "C:/Users/mrustl.KWB/Documents/RProjects/qmra-db.RData"
#save.image(file = file)
#load(file)

table_names <- grep("^tbl", ls(), value = TRUE)

#table_names <- c("tbl_health", "tbl_inflow")

tables <- setNames(lapply(table_names, get), table_names)



for (table_name in table_names) {
  
  kwb.db::hsPutTable(
    path_db, tables[[table_name]], tbl = table_name, overwrite = TRUE
  )
}

}